### ISSUE TITLE
How to choose parameters? 

### ISSUE THREAD

**Issue by grst** [JEFworks-Lab/STalign#19]
In the tutorial, different values for the different sigmas are chosen for the different examples. No rationale is given about the choice of these parameters, except for sigmaM, where the API documentation says:

> As an common example (rule of thumb), you could chose this parameter to be the variance of the pixels in your target image.

Do you have any recommendations how I can find the appropriate parameters for my data? 

--- 

More concretely, I am trying to align an IHC with an ISH image. For testing, I'm focusing on this region if the image, converted to grayscale: 
![image](https://github.com/JEFworks-Lab/STalign/assets/7051479/57e344e1-320d-464a-8a62-ece5e35422b7)

With default parameters, the model doesn't converge, so I suspect that I need to tune the parameters, but I'm unsure where to start. 

Here are the diagnostic plots generated: 

![image](https://github.com/JEFworks-Lab/STalign/assets/7051479/2b2db4a7-4ee2-432e-b1f6-18cd36c05a3b)
![image](https://github.com/JEFworks-Lab/STalign/assets/7051479/75932f92-9b4c-4e26-ab35-309d55fd935d)
![image](https://github.com/JEFworks-Lab/STalign/assets/7051479/15cac24a-b1f0-4794-9afe-de4ea16f957f)

**Comment by kpclifton** [comment]
Hi Gregor,

Thank you for your question about guidance for choosing appropriate parameters for your data. Also it is appreciated that you have included your images and diagnostics plots.

Based on the Weights plot, I believe that for your case the default `sigmaM` value is too high to effectively utilize the Gaussian Mixture Model component of STalign, which labels pixels as background, matching, and artifact tissue based on pixel intensity. If the model is tuned correctly, then in the Weights plot background will be blue, tissue to be matched will be red, and artifacts will be green.

For the `sigmaM`, `sigmaA`, and `sigmaB` values, I would suggest to tune these values to be on the scale of the standard deviation of the pixel intensity of the target image.

To do this, I usually plot a histogram of the pixel intensity values from which I can estimate the sigmas as half the width of the peaks that correspond to background, matching, and artifact tissue.  
In the example below, the background is the peak around 0, the matching tissue is the peak at 0.6, and if there are artifacts they would likely be around 1, and so I would begin by estimating the values as `sigmaB = 0.05`, `sigmaM = 0.3`, and `sigmaA = 0.05`. From my experience, it is not that important that these be precise, but they should at least be the correct order of magnitude for your intensity values.

![Intensity Histogram](https://github.com/JEFworks-Lab/STalign/assets/54687845/7f099b97-a4bb-452d-82db-d9a99b325111)

Additionally, based on your resulting smoothing kernel plot, I would suggest next to begin modifying the spatial smoothness constant `a`. This value has units of length and therefore, it should be tuned relative to the size of the images of interest. Your images seem smaller than the examples in the tutorials so I would suggest to reduce `a` until you get a smoothing kernel plot that has the color variation like the example below. As you tune `a` it is worth keeping in mind that small values of `a` may be overfitting noise and large values of `a` may lead to low accuracy.


![smoothing kernel](https://github.com/JEFworks-Lab/STalign/assets/54687845/1a7af55e-606a-45ab-825f-2574a0192e52)

Hopefully this gives you some direction to begin exploring the parameter space. 

Best,
Kalen

**Comment by grst** [comment]
Hi Kalen, 

thanks for your detailed response! With this information, I could make it *somewhat* work but the result is still not satisfying. Is there anything else you'd suggest to optimize? I also have been wondering about how to choose the learning rate `epV`. The default value is `2000`, however in the tutorial, you selected `10`.

---


I now switched from the b/w image to RGB to get closer to my real use-case:
![image](https://github.com/JEFworks-Lab/STalign/assets/7051479/9080485f-dc34-41cb-9250-78bcf2b1e614)

I chose the first as "source" because it's the more "complete" image (in the right image there's missing a section in the bottom left)

```python
fig, ax = plt.subplots()
ax.hist(J.ravel(), bins=50)
```
![image](https://github.com/JEFworks-Lab/STalign/assets/7051479/5a51a660-f50c-4995-b807-d6035101111b)

Clipping away the peak at the right, I very roughly determined the standard deviation of the "signal peak" as follows:
```python
>>> np.std(J.ravel()[J.ravel() < 0.8])
0.1475
```


Using these parameters
```python
out = STalign.LDDMM(
    [YI, XI],
    I,
    [YJ, XJ],
    J,
    device="cuda:0",
    niter=5000,
    **{
        "sigmaM": 0.15,  # standard deviation of signal peak (tried also 0.3)
        "sigmaB": 0.01,  # standard deviation of background peak 
        "sigmaA": 0.01,  # standard deviation of artifact peak
        "epV": 10,
        "a": 50,
        "muB": torch.tensor([215, 215, 215]) / 255.0,  # "grey" background determined with eyedroper
        "muA": torch.tensor([0, 0, 0]),  # use black as artifact (?)
    },
)
```

I arrive at the following result: 
![image](https://github.com/JEFworks-Lab/STalign/assets/7051479/20eb45d5-0cdf-4333-be2a-034a4a0755b6)
![image](https://github.com/JEFworks-Lab/STalign/assets/7051479/1011313e-78b4-4b73-bf27-0448c89403b1)
![image](https://github.com/JEFworks-Lab/STalign/assets/7051479/9ed2a452-7781-4859-8b9b-1c8cd73c21c4)

**Comment by DruncBread** [comment]
Hi, 

I tried to tune these parameters on my dataset but the result was not good. Would you mind helping me to choose the correct parameters for my dataset? It would be great if you could kindly explain the intuition behind these choice. Thank you! 

[STalign.zip](https://github.com/user-attachments/files/18723261/STalign.zip)
